### ISSUE TITLE
DICE loss

### ISSUE THREAD

**Issue by innat** [keras-team/keras-cv#296]
It is one of the most used loss functions in semantic segmentation tasks. Unfortunately still not available in `keras.loss.*`.  It has been asked many times, [#3611](https://github.com/keras-team/keras/issues/3611), [#13085](https://github.com/keras-team/keras/issues/13085), [#9395](https://github.com/keras-team/keras/issues/9395), [#10890](https://github.com/keras-team/keras/issues/10890).

(Not sure, if it's fit here or `keras.loss.*`)

**Comment by bhack** [comment]
https://github.com/tensorflow/addons/pull/2558#issuecomment-941220370

**Comment by innat** [comment]
Okayyy, someone pushed. I think this function should be available here. It's widely used. 
cc. @gamenerd457

**Comment by gamenerd457** [comment]
Interested

**Comment by bhack** [comment]
@LukeWood If you want this please assign it to @gamenerd457 to port his PR here.

**Comment by innat** [comment]
@LukeWood @qlzh727 
Could u please give some feedback on this and also https://github.com/keras-team/keras-cv/issues/297?

**Comment by qlzh727** [comment]
Yea, I think we can host this here (adding this to keras repo might need more extended API review and requirement). We should put this to keras_cv/losses folder.

**Comment by bhack** [comment]
@gamenerd457 Green light for a PR here.

**Comment by LukeWood** [comment]
yeah seems like a great fit!

**Comment by gamenerd457** [comment]
So should I make a pr here

**Comment by LukeWood** [comment]
> So should I make a pr here

If the loss easily confirms to the Keras loss API (y_true, y_pred) then yes!

**Comment by bhack** [comment]
@LukeWood The intention was to port/refactoring @gamenerd457's Addons PR https://github.com/tensorflow/addons/pull/2558/files

**Comment by innat** [comment]
@gamenerd457 (cc. @bhack @LukeWood @qlzh727 ) 

> If the loss easily confirms to the Keras loss API (y_true, y_pred) then yes!

Some refactoring may be needed, (i.e. subclassing the `tf.keras.losses.Loss` class). I think you can start working on it, either you can move PR from tf-addons or create a new one here. The implementation confirms the keras loss API.

Also please follow the implementation details from [segmentation_models/losses.py](https://github.com/qubvel/segmentation_models/blob/94f624b7029deb463c859efbd92fa26f512b52b8/segmentation_models/losses.py#L51). This reference implementation offers some useful initial parameter (i.e `per_image`, `class_weights` etc), which I found very useful in practice.

**Comment by LukeWood** [comment]
What are some use cases for per_image?  

Also note that in keras we won't support `class_weights`; instead, we will support `sample_weights` in call.

**Comment by innat** [comment]
From [HERE](https://github.com/qubvel/segmentation_models/blob/94f624b7029deb463c859efbd92fa26f512b52b8/segmentation_models/losses.py#L70).

> **per_image**: If ``True`` loss is calculated for each image in batch and then averaged,
        else loss is calculated for the whole batch.

Enabling it to `True` provides a performance boost depending on the task.

**Comment by LukeWood** [comment]
Cool, we will need to include this in the docstring.  Seems like a match!

**Comment by innat** [comment]
@DavidLandup0 

Here are some info that might be helpful.

- about `per_image`, [check 1](https://github.com/keras-team/keras-cv/issues/296#issuecomment-1104167360), [check 2](https://github.com/keras-team/keras-cv/issues/296#issuecomment-1103664354), either investigate or consider [this comment](https://github.com/keras-team/keras-cv/pull/371#discussion_r864179649).
- For [adaptive](https://github.com/keras-team/keras-cv/blob/c298def371872b9f946ab2dd032d5d46f2f03151/keras_cv/losses/dice.py#L51), [generalized](https://github.com/keras-team/keras-cv/blob/c298def371872b9f946ab2dd032d5d46f2f03151/keras_cv/losses/dice.py#L39) dice, [check reference](https://github.com/tensorflow/models/blob/871c4e0a393ef4385534bee55354a5df8aa1ccf4/official/projects/volumetric_models/losses/segmentation_losses.py) it. 
- More stuff regarding dice https://github.com/tensorflow/models/search?q=dice.
- Most popular lib. [segmentation-keras.](https://github.com/qubvel/segmentation_models/blob/master/segmentation_models/losses.py)
